File: src/services/renderers/speaker3D.ts
import { RenderContext } from './types';

let _cachedH = 0;
let _barGrad: CanvasGradient | null = null;

export function renderSpeaker3D(r: RenderContext) {
  const { ctx: c, width, height, config, freqData, bassEnergy: be, beatStrength: bs } = r;
  const sensitivity = config.reactivity.sensitivity;
  const theme = config.theme;

  const centerX = width / 2;
  const centerY = height / 2;

  const baseRadius = Math.min(width, height) * 0.27;
  const bassPulse = 1.0 + be * 0.12 + bs * 0.08;
  const speakerR = baseRadius * bassPulse;

  function hexToRgb(hex: string): [number, number, number] {
    const h = hex.replace('#', '');
    return [parseInt(h.slice(0, 2), 16), parseInt(h.slice(2, 4), 16), parseInt(h.slice(4, 6), 16)];
  }
  const [pR, pG, pB] = hexToRgb(theme.primaryColor);
  const [sR, sG, sB] = hexToRgb(theme.secondaryColor);
  const [aR, aG, aB] = hexToRgb(theme.accentColor);

  if (!_barGrad || _cachedH !== height) {
    _cachedH = height;
    _barGrad = c.createLinearGradient(0, 0, 0, height);
    _barGrad.addColorStop(0, `rgba(${pR}, ${pG}, ${pB}, 0.85)`);
    _barGrad.addColorStop(0.3, `rgba(${sR}, ${sG}, ${sB}, 0.95)`);
    _barGrad.addColorStop(0.6, `rgba(${aR}, ${aG}, ${aB}, 0.98)`);
    _barGrad.addColorStop(0.85, `rgba(${sR}, ${sG}, ${sB}, 0.95)`);
    _barGrad.addColorStop(1, `rgba(${pR}, ${pG}, ${pB}, 0.85)`);
  }

  c.save();

  // --- PASS 1: BACKGROUND HIGH-DENSITY FIERY SPECTRUM BARS (LEFT & RIGHT WINGS) ---
  const halfBars = 48;
  const step = Math.max(1, Math.floor((freqData.length * 0.7) / halfBars));

  const leftStart = width * 0.02;
  const leftEnd = Math.max(leftStart + 20, centerX - speakerR * 0.85);
  const leftWidth = leftEnd - leftStart;

  const rightStart = Math.min(width * 0.98 - 20, centerX + speakerR * 0.85);
  const rightEnd = width * 0.98;
  const rightWidth = rightEnd - rightStart;

  const barW = Math.max(2.5, (leftWidth / halfBars) - 2.5);

  // Laser baseline passing behind speaker
  c.shadowBlur = 20 + be * 20;
  c.shadowColor = theme.glowColor || '#FF4400';
  c.strokeStyle = theme.primaryColor || '#FF7700';
  c.lineWidth = 2.2;
  c.beginPath();
  c.moveTo(0, centerY);
  c.lineTo(width, centerY);
  c.stroke();

  c.shadowBlur = 15;
  c.shadowColor = theme.glowColor || '#FF5500';

  for (let i = 0; i < halfBars; i++) {
    let sum = 0;
    for (let j = 0; j < step; j++) {
      sum += freqData[i * step + j] || 0;
    }
    const val = (sum / step / 255) * sensitivity;
    if (val < 0.01) continue;

    const barH = val * height * 0.36;
    const topY = centerY - barH;
    const botY = centerY + barH * 0.82;

    const xLeft = leftEnd - (i / (halfBars - 1)) * leftWidth - barW;
    const xRight = rightStart + (i / (halfBars - 1)) * rightWidth;

    c.fillStyle = _barGrad;
    c.fillRect(xLeft, topY, barW, barH * 1.82);
    c.fillRect(xRight, topY, barW, barH * 1.82);

    c.fillStyle = theme.accentColor;
    c.fillRect(xLeft - 0.5, topY - 1.5, barW + 1, 1.5);
    c.fillRect(xLeft - 0.5, botY, barW + 1, 1.5);
    c.fillRect(xRight - 0.5, topY - 1.5, barW + 1, 1.5);
    c.fillRect(xRight - 0.5, botY, barW + 1, 1.5);
  }

  // --- PASS 2: SPEAKER AURA & LENS FLARES ---
  const glowR = speakerR * 1.4;
  const backGlow = c.createRadialGradient(centerX, centerY, speakerR * 0.5, centerX, centerY, glowR);
  backGlow.addColorStop(0, `rgba(${aR}, ${aG}, ${aB}, 0.85)`);
  backGlow.addColorStop(0.5, `rgba(${sR}, ${sG}, ${sB}, 0.4)`);
  backGlow.addColorStop(1, 'rgba(0, 0, 0, 0)');

  c.shadowBlur = 45 + be * 35;
  c.shadowColor = theme.glowColor;
  c.fillStyle = backGlow;
  c.beginPath();
  c.arc(centerX, centerY, glowR, 0, Math.PI * 2);
  c.fill();

  const flareX = (dx: number) => {
    const fx = centerX + dx;
    const fg = c.createRadialGradient(fx, centerY, 0, fx, centerY, speakerR * 0.45);
    fg.addColorStop(0, 'rgba(255, 255, 255, 0.95)');
    fg.addColorStop(0.2, `rgba(${pR}, ${pG}, ${pB}, 0.85)`);
    fg.addColorStop(0.6, `rgba(${sR}, ${sG}, ${sB}, 0.3)`);
    fg.addColorStop(1, 'rgba(0, 0, 0, 0)');
    c.fillStyle = fg;
    c.beginPath();
    c.arc(fx, centerY, speakerR * 0.45, 0, Math.PI * 2);
    c.fill();
  };
  flareX(-speakerR * 0.96);
  flareX(speakerR * 0.96);

  // --- PASS 3: OUTSIDE CHROME METALLIC RIM ---
  const outerRimR = speakerR;
  const innerRimR = speakerR * 0.88;

  // 3D Metallic Gradient
  const metallicGrad = c.createLinearGradient(
    centerX - outerRimR, centerY - outerRimR,
    centerX + outerRimR, centerY + outerRimR
  );
  metallicGrad.addColorStop(0.0, '#FFFFFF');
  metallicGrad.addColorStop(0.15, '#8E8E93');
  metallicGrad.addColorStop(0.35, '#2C2C2E');
  metallicGrad.addColorStop(0.55, '#D1D1D6');
  metallicGrad.addColorStop(0.75, '#48484A');
  metallicGrad.addColorStop(1.0, '#E5E5EA');

  c.shadowBlur = 18;
  c.shadowColor = '#000000';
  c.fillStyle = metallicGrad;
  c.beginPath();
  c.arc(centerX, centerY, outerRimR, 0, Math.PI * 2);
  c.arc(centerX, centerY, innerRimR, 0, Math.PI * 2, true);
  c.closePath();
  c.fill();

  // Highlight Ring Edges
  c.strokeStyle = 'rgba(255, 255, 255, 0.5)';
  c.lineWidth = 1.5;
  c.beginPath();
  c.arc(centerX, centerY, outerRimR - 1, 0, Math.PI * 2);
  c.stroke();

  c.strokeStyle = 'rgba(0, 0, 0, 0.6)';
  c.lineWidth = 1.5;
  c.beginPath();
  c.arc(centerX, centerY, innerRimR + 1, 0, Math.PI * 2);
  c.stroke();

  // Screws at 12, 3, 6, 9 o'clock
  const boltRadius = (outerRimR + innerRimR) / 2;
  for (let a = 0; a < 4; a++) {
    const angle = a * (Math.PI / 2);
    const bx = centerX + Math.cos(angle) * boltRadius;
    const by = centerY + Math.sin(angle) * boltRadius;

    c.shadowBlur = 4;
    c.shadowColor = '#000000';
    c.fillStyle = '#E5E5EA';
    c.beginPath();
    c.arc(bx, by, 3.8, 0, Math.PI * 2);
    c.fill();

    // Screw center dot & slot line
    c.strokeStyle = '#1C1C1E';
    c.lineWidth = 1.2;
    c.beginPath();
    c.moveTo(bx - 2, by);
    c.lineTo(bx + 2, by);
    c.stroke();
  }

  // --- PASS 4: FLEXIBLE RUBBER SURROUND RING ---
  const surroundOuterR = innerRimR;
  const surroundInnerR = speakerR * 0.74;

  const rubberGrad = c.createRadialGradient(
    centerX, centerY, surroundInnerR,
    centerX, centerY, surroundOuterR
  );
  rubberGrad.addColorStop(0, '#1C1C1E');
  rubberGrad.addColorStop(0.5, '#3A3A3C');
  rubberGrad.addColorStop(1, '#0C0C0E');

  c.fillStyle = rubberGrad;
  c.beginPath();
  c.arc(centerX, centerY, surroundOuterR, 0, Math.PI * 2);
  c.arc(centerX, centerY, surroundInnerR, 0, Math.PI * 2, true);
  c.closePath();
  c.fill();

  // --- PASS 5: POLKA-DOT MESH CONE DIAPHRAGM ---
  const coneOuterR = surroundInnerR;
  const coneInnerR = speakerR * 0.30;

  const coneGrad = c.createRadialGradient(
    centerX - coneOuterR * 0.25, centerY - coneOuterR * 0.25, coneInnerR * 0.4,
    centerX, centerY, coneOuterR
  );
  coneGrad.addColorStop(0, '#48484A');
  coneGrad.addColorStop(0.5, '#2C2C2E');
  coneGrad.addColorStop(1, '#1C1C1E');

  c.fillStyle = coneGrad;
  c.beginPath();
  c.arc(centerX, centerY, coneOuterR, 0, Math.PI * 2);
  c.fill();

  c.fillStyle = 'rgba(255, 255, 255, 0.08)';
  const gridSpacing = 12;
  const dotR = 1.8;
  const minD = coneInnerR * 0.9;
  const maxD = coneOuterR * 0.98;
  const minD2 = minD * minD;
  const maxD2 = maxD * maxD;
  for (let gy = centerY - coneOuterR; gy <= centerY + coneOuterR; gy += gridSpacing) {
    const rowOffset = Math.floor((gy - centerY) / gridSpacing) % 2 === 0 ? 0 : gridSpacing * 0.5;
    for (let gx = centerX - coneOuterR; gx <= centerX + coneOuterR; gx += gridSpacing) {
      const xPos = gx + rowOffset;
      const dx = xPos - centerX;
      const dy = gy - centerY;
      const dist2 = dx * dx + dy * dy;
      if (dist2 >= minD2 && dist2 <= maxD2) {
        c.beginPath();
        c.arc(xPos, gy, dotR, 0, Math.PI * 2);
        c.fill();
      }
    }
  }

  // --- PASS 6: 3D DUST CAP CONE WITH GLOSSY CRESCENT GLARE ---
  const dustCapR = coneInnerR * (1.0 + be * 0.06);

  const dustCapGrad = c.createRadialGradient(
    centerX - dustCapR * 0.3, centerY - dustCapR * 0.3, 0,
    centerX, centerY, dustCapR
  );
  dustCapGrad.addColorStop(0, '#636366');
  dustCapGrad.addColorStop(0.4, '#3A3A3C');
  dustCapGrad.addColorStop(0.85, '#1C1C1E');
  dustCapGrad.addColorStop(1, '#0C0C0E');

  c.shadowBlur = 14 + be * 10;
  c.shadowColor = '#000000';
  c.fillStyle = dustCapGrad;
  c.beginPath();
  c.arc(centerX, centerY, dustCapR, 0, Math.PI * 2);
  c.fill();

  // Glossy Crescent Reflection Arc
  c.fillStyle = 'rgba(255, 255, 255, 0.35)';
  c.beginPath();
  c.arc(centerX - dustCapR * 0.15, centerY - dustCapR * 0.15, dustCapR * 0.65, Math.PI * 1.0, Math.PI * 1.85);
  c.arc(centerX - dustCapR * 0.15, centerY - dustCapR * 0.15, dustCapR * 0.45, Math.PI * 1.85, Math.PI * 1.0, true);
  c.closePath();
  c.fill();

  c.restore();
}

File: src/renderers/styles/speaker_3d.rs
//! Speaker 3D style renderer (`speaker3D`).

use std::f32::consts::TAU;

use crate::gpu2d::{Color, Fill, GpuCanvas};
use crate::renderers::helpers::bin_sum;
use crate::renderers::RenderContext;

pub fn render(c: &mut GpuCanvas, ctx: &mut RenderContext) {
  let width = ctx.width;
  let height = ctx.height;
  let theme = &ctx.config.theme;
  let p = crate::renderers::theme_primary(theme);
  let s = crate::renderers::theme_secondary(theme);
  let a = crate::renderers::theme_accent(theme);
  let glow = crate::renderers::theme_glow(theme);
  let sensitivity = ctx.config.reactivity.sensitivity;
  let be = ctx.bass_energy;
  let bs = ctx.beat_strength;
  let freq = ctx.freq_data;

  let center_x = width / 2.0;
  let center_y = height / 2.0;
  let base_radius = width.min(height) * 0.27;
  let bass_pulse = 1.0 + be * 0.12 + bs * 0.08;
  let speaker_r = base_radius * bass_pulse;

  let bar_grad = Fill::linear_gradient(0.0, 0.0, 0.0, height, &[
    (0.0, p.with_alpha(0.85)),
    (0.3, s.with_alpha(0.95)),
    (0.6, a.with_alpha(0.98)),
    (0.85, s.with_alpha(0.95)),
    (1.0, p.with_alpha(0.85)),
  ]);

  c.save();
  c.set_shadow(Color::TRANSPARENT, 0.0);

  let half_bars = 48;
  let step = ((freq.len() as f32 * 0.7) as usize / half_bars).max(1);
  let left_start = width * 0.02;
  let left_end = (left_start + 20.0).max(center_x - speaker_r * 0.85);
  let left_width = left_end - left_start;
  let right_start = (width * 0.98 - 20.0).min(center_x + speaker_r * 0.85);
  let right_end = width * 0.98;
  let right_width = right_end - right_start;
  let bar_w = ((left_width / half_bars as f32) - 2.5).max(2.5);

  c.set_shadow(glow, 20.0 + be * 20.0);
  c.set_stroke(Fill::Solid(p));
  c.set_line_width(2.2);
  c.stroke_line(0.0, center_y, width, center_y);
  c.set_shadow(glow, 15.0);

  for i in 0..half_bars {
    let val = bin_sum(freq, step, i) * sensitivity;
    if val < 0.01 {
      continue;
    }
    let bar_h = val * height * 0.36;
    let top_y = center_y - bar_h;
    let bot_y = center_y + bar_h * 0.82;
    let f = i as f32 / (half_bars - 1) as f32;
    let x_left = left_end - f * left_width - bar_w;
    let x_right = right_start + f * right_width;
    c.set_fill(bar_grad.clone());
    c.fill_rect(x_left, top_y, bar_w, bar_h * 1.82);
    c.fill_rect(x_right, top_y, bar_w, bar_h * 1.82);
    c.set_fill(Fill::Solid(a));
    c.fill_rect(x_left - 0.5, top_y - 1.5, bar_w + 1.0, 1.5);
    c.fill_rect(x_left - 0.5, bot_y, bar_w + 1.0, 1.5);
    c.fill_rect(x_right - 0.5, top_y - 1.5, bar_w + 1.0, 1.5);
    c.fill_rect(x_right - 0.5, bot_y, bar_w + 1.0, 1.5);
  }

  let glow_r = speaker_r * 1.4;
  let back_glow = Fill::radial_gradient(
    center_x,
    center_y,
    speaker_r * 0.5,
    center_x,
    center_y,
    glow_r,
    &[
      (0.0, a.with_alpha(0.85)),
      (0.5, s.with_alpha(0.4)),
      (1.0, Color::TRANSPARENT),
    ],
  );
  c.set_shadow(glow, 45.0 + be * 35.0);
  c.set_fill(back_glow);
  c.fill_circle(center_x, center_y, glow_r);

  let flare = |c: &mut GpuCanvas, fx: f32| {
    let fg = Fill::radial_gradient(
      fx,
      center_y,
      0.0,
      fx,
      center_y,
      speaker_r * 0.45,
      &[
        (0.0, Color::rgba(1.0, 1.0, 1.0, 0.95)),
        (0.2, p.with_alpha(0.85)),
        (0.6, s.with_alpha(0.3)),
        (1.0, Color::TRANSPARENT),
      ],
    );
    c.set_fill(fg);
    c.fill_circle(fx, center_y, speaker_r * 0.45);
  };
  flare(c, center_x - speaker_r * 0.96);
  flare(c, center_x + speaker_r * 0.96);

  let outer_rim = speaker_r;
  let inner_rim = speaker_r * 0.88;
  let metallic = Fill::linear_gradient(
    center_x - outer_rim,
    center_y - outer_rim,
    center_x + outer_rim,
    center_y + outer_rim,
    &[
      (0.0, Color::hex("#FFFFFF")),
      (0.15, Color::hex("#8E8E93")),
      (0.35, Color::hex("#2C2C2E")),
      (0.55, Color::hex("#D1D1D6")),
      (0.75, Color::hex("#48484A")),
      (1.0, Color::hex("#E5E5EA")),
    ],
  );
  c.set_shadow(Color::hex("#000000"), 18.0);
  c.set_fill(metallic);
  c.fill_ring(center_x, center_y, outer_rim, inner_rim);

  c.set_shadow(Color::TRANSPARENT, 0.0);
  c.set_stroke(Fill::Solid(Color::rgba(1.0, 1.0, 1.0, 0.5)));
  c.set_line_width(1.5);
  c.stroke_circle(center_x, center_y, outer_rim - 1.0);
  c.set_stroke(Fill::Solid(Color::rgba(0.0, 0.0, 0.0, 0.6)));
  c.stroke_circle(center_x, center_y, inner_rim + 1.0);

  let bolt_radius = (outer_rim + inner_rim) / 2.0;
  for k in 0..4 {
    let angle = k as f32 * TAU / 4.0;
    let bx = center_x + angle.cos() * bolt_radius;
    let by = center_y + angle.sin() * bolt_radius;
    c.set_shadow(Color::hex("#000000"), 4.0);
    c.set_fill(Fill::Solid(Color::hex("#E5E5EA")));
    c.fill_circle(bx, by, 3.8);
    c.set_shadow(Color::TRANSPARENT, 0.0);
    c.set_stroke(Fill::Solid(Color::hex("#1C1C1E")));
    c.set_line_width(1.2);
    c.stroke_line(bx - 2.0, by, bx + 2.0, by);
  }

  let surround_outer = inner_rim;
  let surround_inner = speaker_r * 0.74;
  let rubber = Fill::radial_gradient(
    center_x,
    center_y,
    surround_inner,
    center_x,
    center_y,
    surround_outer,
    &[
      (0.0, Color::hex("#1C1C1E")),
      (0.5, Color::hex("#3A3A3C")),
      (1.0, Color::hex("#0C0C0E")),
    ],
  );
  c.set_fill(rubber);
  c.fill_ring(center_x, center_y, surround_outer, surround_inner);

  let cone_outer = surround_inner;
  let cone_inner = speaker_r * 0.30;
  let cone = Fill::radial_gradient(
    center_x - cone_outer * 0.25,
    center_y - cone_outer * 0.25,
    cone_inner * 0.4,
    center_x,
    center_y,
    cone_outer,
    &[
      (0.0, Color::hex("#48484A")),
      (0.5, Color::hex("#2C2C2E")),
      (1.0, Color::hex("#1C1C1E")),
    ],
  );
  c.set_fill(cone);
  c.fill_circle(center_x, center_y, cone_outer);

  c.set_shadow(Color::TRANSPARENT, 0.0);
  c.set_fill(Fill::Solid(Color::rgba(1.0, 1.0, 1.0, 0.08)));
  let grid = 12.0f32;
  let dot_r = 1.8;
  let min_d2 = (cone_inner * 0.9) * (cone_inner * 0.9);
  let max_d2 = (cone_outer * 0.98) * (cone_outer * 0.98);
  let mut gy = center_y - cone_outer;
  while gy <= center_y + cone_outer {
    // TS: `Math.floor((gy - centerY) / gridSpacing) % 2 === 0 ? 0 : gridSpacing * 0.5`
    // (speaker3D.ts:225). The parity comes from the absolute gy, NOT a row
    // counter — a plain counter phase-inverts the whole dot mesh whenever
    // `ceil(coneOuter / grid)` is odd.
    let row_offset = if (((gy - center_y) / grid).floor() as i32) % 2 == 0 { 0.0 } else { grid * 0.5 };
    let mut gx = center_x - cone_outer;
    while gx <= center_x + cone_outer {
      let xp = gx + row_offset;
      let dxp = xp - center_x;
      let dyp = gy - center_y;
      let d2 = dxp * dxp + dyp * dyp;
      if d2 >= min_d2 && d2 <= max_d2 {
        c.fill_circle(xp, gy, dot_r);
      }
      gx += grid;
    }
    gy += grid;
  }

  let dust_r = cone_inner * (1.0 + be * 0.06);
  let dust = Fill::radial_gradient(
    center_x - dust_r * 0.3,
    center_y - dust_r * 0.3,
    0.0,
    center_x,
    center_y,
    dust_r,
    &[
      (0.0, Color::hex("#636366")),
      (0.4, Color::hex("#3A3A3C")),
      (0.85, Color::hex("#1C1C1E")),
      (1.0, Color::hex("#0C0C0E")),
    ],
  );
  c.set_shadow(Color::hex("#000000"), 14.0 + be * 10.0);
  c.set_fill(dust);
  c.fill_circle(center_x, center_y, dust_r);
  c.set_shadow(Color::TRANSPARENT, 0.0);
  c.set_fill(Fill::Solid(Color::rgba(1.0, 1.0, 1.0, 0.35)));
  c.fill_ring_arc(
    center_x - dust_r * 0.15,
    center_y - dust_r * 0.15,
    dust_r * 0.65,
    dust_r * 0.45,
    TAU * 0.5,
    TAU * 0.925,
  );

  c.restore();
}
